library(DESeq2)

# load data generated by DEseq2
dds <- readRDS("./data/DE_analysis/DeSeq2_dds.RDS")
vsd <- vst(dds, blind = FALSE)


# ---------------------------------
########### make PCA plots
# ---------------------------------
# change the label here
pdf("./plots/transcriptomic/DeSeq2/PCA/pca.pdf", width = 4, height = 4)
plotPCA(vsd, intgroup = c("condition")) +
  geom_point(size = 5) +
  theme_bw() + 
  scale_x_continuous(limits = c(-30, 20)) +
  scale_y_continuous(limits = c(-30, 20)) +
  theme(legend.position = "none") 
dev.off()

# ---------------------------------
############## make volcano plot
# ---------------------------------
# add gene names
gtf <- readRDS("./data/R_data/c_elegans_gene_gtf.RDS")
df_data$gene_name <- gtf[rownames(df_data), "gene_name"]

# prepare input data
groups_adjust_p <- colnames(df_data[,grepl("(.*)_adjustP", colnames(df_data))])
groups_log_fc <- colnames(df_data[,grepl("(.*)_logFC", colnames(df_data))])
groups_pvalue <- colnames(df_data[,grepl("(.*)_Pvalue", colnames(df_data))])

# make plots
mapply(x = groups_log_fc,
       y = groups_adjust_p,
       z = groups_pvalue, FUN = function(x, y, z){
         df_genes_data <- df_data[!is.na(df_data[, y]), c(x, y, z, "gene_name")]
         colnames(df_genes_data) <- c("logFC","adjustP", "pValue", "gene_name")
         
         df_genes_data$Significance <-  "Not significant"
         df_genes_data$Significance[df_genes_data$adjustP < 0.05 &
                                      df_genes_data$logFC >= 1] <- "up regulated"
         df_genes_data$Significance[df_genes_data$adjustP < 0.05 &
                                      df_genes_data$logFC <= -1] <- "down regulated"
         title <- gsub("(.*)_logFC", "\\1", x)
         
         pdf(file = paste0("./plots/transcriptomic/DeSeq2/volcano/", title, ".pdf"), width = 5, height = 4)
         print(ggplot(df_genes_data, aes(x = logFC, y = -log10(pValue), color = Significance)) +
           geom_point(size = 1.5) +
           scale_color_manual(values=c("Not significant" = "darkgrey", 
                                       "up regulated" = "red", 
                                       "down regulated" = "blue")) +  # geom_text_repel(data=filter(results_cleaned, adj.p.val<cutoff), aes(label=Name)) +
           labs(x="log2(fold change)", y="-log10(P.value)") +
           scale_y_continuous(expand = c(0,0)) +
           theme_bw() +
           ggtitle(title))
         
         dev.off()
         }, SIMPLIFY = FALSE)

